File: old_contents/boj_codes/Python3/1991.py
#https://www.acmicpc.net/problem/1991
# 20200819

import sys

class Node:
    def __init__(self, item, left, right, parent):
        self.item = item
        self.left = None
        self.right = None
        self.parent = parent

        if left != ".":
            self.left = Node(left, ".", ".", self)

        if right != ".":
            self.right = Node(right, ".", ".", self)

    def is_left(self):
        if self.left is None:
            return False
        return True

    def is_right(self):
        if self.right is None:
            return False
        return True

class Tree:
    def __init__(self):
        self.Nnode = 0
        self.root = None

    def input_nodes(self, item, left, right):
        if self.Nnode == 0:
            self.root = Node(item, left, right, None)

        else:
            node = self.traverse(self.root, item)
            if node is None:
                raise NameError

            if node.parent.is_left() and node.parent.left.item == item:
                node.parent.left = Node(item, left, right, node.parent)

            else:
                node.parent.right = Node(item, left, right, node.parent)


        self.Nnode += 1


    
    def traverse(self, node, item):
        if node.item == item:
            return node

        if node.is_left():
            result_left = self.traverse(node.left, item)
            if result_left is not None:
                return result_left

        if node.is_right():
            result_right = self.traverse(node.right, item)
            if result_right is not None:
                return result_right

        return None


    def preorder(self, node):
        print(node.item, end = '')

        if node.is_left():
            self.preorder(node.left)

        if node.is_right():
            self.preorder(node.right)

    def inorder(self, node):

        if node.is_left():
            self.inorder(node.left)

        print(node.item, end = '')

        if node.is_right():
            self.inorder(node.right)

    def postorder(self, node):

        if node.is_left():
            self.postorder(node.left)

        if node.is_right():
            self.postorder(node.right)

        print(node.item, end= '')


if __name__ == "__main__":
    Nnode = int(sys.stdin.readline())
    tree = Tree()

    for i in range(Nnode):
        item, left, right = tuple(sys.stdin.readline().split())
        tree.input_nodes(item, left, right)

    tree.preorder(tree.root)
    print()
    tree.inorder(tree.root)
    print()
    tree.postorder(tree.root)
    print()

File: old_contents/boj_codes/C++17/1991.cpp
// https://www.acmicpc.net/problem/1991
// First Written : 20210105
// Last Modified : 20210105

#include <cstdio>
#include <vector>

class Node{
private:
    char name;
public:
    Node* left;
    Node* right;
    Node* parent;

    Node(char name){
        this->left = nullptr;
        this->right = nullptr;
        this->parent = nullptr;
        this->name = name;
    }

    void add_left_c(Node* left_c){
        this->left = left_c;
        left_c->parent = this;
    }

    void add_right_c(Node* right_c){
        this->right = right_c;
    }

    inline bool is_left(){
        if(this->left)
            return true;
        
        else
            return false;
    }

    inline bool is_right(){
        if(this->right)
            return true;
        
        else
            return false;
    }

    char get_name(){
        return this->name;
    }

};

class Graph{
private:
    Node* root;

public:
    Graph(){
        Node * root_node = new Node('A');
        this->root = root_node;
    }

    Node* traverse(Node* root_node, char word){
        if(root_node->get_name() == word){
           return root_node;
        }
        else{
            Node* new_nodep;
            if(root_node->is_left()){
                new_nodep = traverse(root_node->left, word);
                if(new_nodep)
                    return new_nodep;
           }

            if(root_node->is_right()){
                new_nodep = traverse(root_node->right, word);
                if(new_nodep)
                    return new_nodep;
            }

            return nullptr;
       }
    }

    void add_item(char parent, char left, char right){
        Node* node = traverse(this->root, parent);

        if(!node){
            printf("fatal error!, %c\n", parent);
            return;
        }

        if(left != '.'){
            Node* new_lnode = new Node(left);
            node->left = new_lnode;
            new_lnode->parent = node;
        }

        if(right != '.'){
            Node* new_rnode = new Node(right);
            node->right = new_rnode;
            new_rnode->parent = node;
        }


    }

    void preorder(Node* root_node){
        printf("%c", root_node->get_name());

        if(root_node->is_left())
            preorder(root_node->left);
        
        if(root_node->is_right())
            preorder(root_node->right);
    }

    void inorder(Node* root_node){
        if(root_node->is_left())
            inorder(root_node->left);
        
        printf("%c", root_node->get_name());

        if(root_node->is_right())
            inorder(root_node->right);

    }

    void postorder(Node* root_node){
        if(root_node->is_left())
            postorder(root_node->left);

        if(root_node->is_right())
            postorder(root_node->right);

        printf("%c", root_node->get_name());
    }

    Node* get_root(){
        return this->root;
    }
};


int main(){
    int N;
    char parent, left, right;
    scanf("%d", &N);

    Graph* graph = new Graph();
    while(N--){
        scanf(" %c %c %c", &parent, &left, &right);
        graph->add_item(parent, left, right);
    }
    
    graph->preorder(graph->get_root());
    printf("\n");
    graph->inorder(graph->get_root());
    printf("\n");
    graph->postorder(graph->get_root());

    delete graph;

    return 0;
}